pytket/extensions/aqt/multi_zone_architecture/circuit/multizone_circuit.py

```diff
@@ -0,0 +1,345 @@
+import itertools
+import logging
+from copy import deepcopy
+from dataclasses import dataclass, field
+from enum import Enum
+from typing import Any, cast
+from sympy import symbols
+
+from pytket.circuit import Circuit, OpType, CustomGateDef  # type: ignore
+
+from ..architecture import (
+    MultiZoneArchitecture,
+    EdgeType,
+    source_edge_type,
+    target_edge_type,
+)
+from ..macro_architechture_graph import (
+    MultiZoneMacroArch,
+    empty_macro_arch_from_backend,
+    ZoneId,
+)
+
+
+class QubitPlacementError(Exception):
+    pass
+
+
+class MoveError(Exception):
+    pass
+
+
+class AcrossZoneOperationError(Exception):
+    pass
+
+
+class VirtualZonePosition(Enum):
+    VirtualLeft = 0
+    VirtualRight = 1
+
+
+dz, se, te = symbols("destination_zone source_edge target_edge")
+
+move_def_circ = Circuit(1)
+move_def_circ.add_barrier([0])
+move_gate = CustomGateDef("MOVE", move_def_circ, [dz])
+
+shuttle_def_circ = Circuit(1)
+shuttle_gate = CustomGateDef("SHUTTLE", shuttle_def_circ, [dz, se, te])
+
+swap_def_circ = Circuit(2)
+swap_gate = CustomGateDef("PSWAP", swap_def_circ, [])
+
+
+@dataclass
+class SwapWithinZone:
+    qubit_0: int
+    qubit_1: int
+
+    def __str__(self) -> str:
+        return f"{self.qubit_0}: {self.qubit_1}"
+
+    def append_to_circuit(self, circuit: "MultiZoneCircuit") -> None:
+        circuit.add_custom_gate(swap_gate, [], [self.qubit_0, self.qubit_1])
+
+
+@dataclass
+class Shuttle:
+    qubit: int
+    zone: int
+
+    source_edge: EdgeType
+    target_edge: EdgeType
+
+    source_edge_int_encoding: int = field(init=False)
+    target_edge_int_encoding: int = field(init=False)
+
+    def __post_init__(self) -> None:
+        """
+        Encode left and right edges as negative or non-negative
+         number for use as tket custom op parameters
+        Yes, this is a hack
+        """
+        self.source_edge_int_encoding = -1 if self.source_edge == EdgeType.Left else 1
+        self.target_edge_int_encoding = -1 if self.target_edge == EdgeType.Left else 1
+
+    def __str__(self) -> str:
+        return f"{self.qubit}: {self.zone}"
+
+    def append_to_circuit(self, circuit: "MultiZoneCircuit") -> None:
+        circuit.add_custom_gate(
+            shuttle_gate,
+            [self.zone, self.source_edge_int_encoding, self.target_edge_int_encoding],
+            [self.qubit],
+        )
+
+
+@dataclass
+class Init:
+    qubit: int
+    zone: int
+
+
+MZAOperation = SwapWithinZone | Shuttle
+
+
+def _swap_left_to_right_through_list(
+    qubit: int, qubit_list: list[int]
+) -> list[MZAOperation]:
+    return [SwapWithinZone(qubit, swap_qubit) for swap_qubit in qubit_list]
+
+
+def _swap_right_to_left_through_list(
+    qubit: int, qubit_list: list[int]
+) -> list[MZAOperation]:
+    return [SwapWithinZone(swap_qubit, qubit) for swap_qubit in reversed(qubit_list)]
+
+
+def _move_from_zone_position_to_connected_zone_edge(
```

Comment:
Can you maybe add a short explaining comment for the hidden functions as well for the developer working on this?

Reply:
Yes sure. I think a lot of this is going to be moved by me to the tket repo pretty soon. That's why I didn't bother documenting it too well. But I think you are right that a little documentation would be good